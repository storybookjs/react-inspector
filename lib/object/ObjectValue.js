'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _typeof2 = require('babel-runtime/helpers/typeof');

var _typeof3 = _interopRequireDefault(_typeof2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createStyles = require('../styles/createStyles');

var _createStyles2 = _interopRequireDefault(_createStyles);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * A short description of the object values.
 * Can be used to render tree node in ObjectInspector
 * or render objects in TableInspector.
 */
var ObjectValue = function ObjectValue(_ref, _ref2) {
  var object = _ref.object;
  var theme = _ref2.theme;

  var styles = (0, _createStyles2.default)('ObjectValue', theme);

  switch (typeof object === 'undefined' ? 'undefined' : (0, _typeof3.default)(object)) {
    case 'number':
      return _react2.default.createElement(
        'span',
        { style: styles.objectValueNumber },
        object
      );
    case 'string':
      return _react2.default.createElement(
        'span',
        { style: styles.objectValueString },
        '"',
        object,
        '"'
      );
    case 'boolean':
      return _react2.default.createElement(
        'span',
        { style: styles.objectValueBoolean },
        String(object)
      );
    case 'undefined':
      return _react2.default.createElement(
        'span',
        { style: styles.objectValueUndefined },
        'undefined'
      );
    case 'object':
      if (object === null) {
        return _react2.default.createElement(
          'span',
          { style: styles.objectValueNull },
          'null'
        );
      }
      if (object instanceof Date) {
        return _react2.default.createElement(
          'span',
          null,
          object.toString()
        );
      }
      if (object instanceof RegExp) {
        return _react2.default.createElement(
          'span',
          { style: styles.objectValueRegExp },
          object.toString()
        );
      }
      if (Array.isArray(object)) {
        return _react2.default.createElement(
          'span',
          null,
          'Array[' + object.length + ']'
        );
      }
      return _react2.default.createElement(
        'span',
        null,
        object.constructor.name
      );
    case 'function':
      return _react2.default.createElement(
        'span',
        null,
        _react2.default.createElement(
          'span',
          { style: styles.objectValueFunctionKeyword },
          'function'
        ),
        _react2.default.createElement(
          'span',
          { style: styles.objectValueFunctionName },
          '\xA0',
          object.name,
          '()'
        )
      );
    case 'symbol':
      return _react2.default.createElement(
        'span',
        { style: styles.objectValueSymbol },
        object.toString()
      );
    default:
      return _react2.default.createElement('span', null);
  }
};

ObjectValue.propTypes = {
  /** the object to describe */
  object: _react.PropTypes.any
};

ObjectValue.contextTypes = {
  theme: _react.PropTypes.oneOfType([_react.PropTypes.string, _react.PropTypes.object])
};

exports.default = ObjectValue;