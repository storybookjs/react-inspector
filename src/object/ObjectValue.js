import React from 'react';
import PropTypes from 'prop-types';
import createStyles from '../styles/createStyles';

/**
 * A short description of the object values.
 * Can be used to render tree node in ObjectInspector
 * or render objects in TableInspector.
 */
const ObjectValue = ({ object, styles }, { theme }) => {
  const themeStyles = createStyles('ObjectValue', theme);

  const mkStyle = key => ({ ...themeStyles[key], ...styles });

  switch (typeof object) {
    case 'number':
      return (
        <span style={mkStyle('objectValueNumber')}>
          {object}
        </span>
      );
    case 'string':
      return (
        <span style={mkStyle('objectValueString')}>
          "{object}"
        </span>
      );
    case 'boolean':
      return (
        <span style={mkStyle('objectValueBoolean')}>
          {String(object)}
        </span>
      );
    case 'undefined':
      return <span style={mkStyle('objectValueUndefined')}>undefined</span>;
    case 'object':
      if (object === null) {
        return <span style={mkStyle('objectValueNull')}>null</span>;
      }
      if (object instanceof Date) {
        return (
          <span>
            {object.toString()}
          </span>
        );
      }
      if (object instanceof RegExp) {
        return (
          <span style={mkStyle('objectValueRegExp')}>
            {object.toString()}
          </span>
        );
      }
      if (Array.isArray(object)) {
        return <span>{`Array[${object.length}]`}</span>;
      }
      return (
        <span>
          {object.constructor.name}
        </span>
      );
    case 'function':
      return (
        <span>
          <span style={mkStyle('objectValueFunctionKeyword')}>function</span>
          <span style={mkStyle('objectValueFunctionName')}>
            &nbsp;{object.name}()
          </span>
        </span>
      );
    case 'symbol':
      return (
        <span style={mkStyle('objectValueSymbol')}>
          {object.toString()}
        </span>
      );
    default:
      return <span />;
  }
};

ObjectValue.propTypes = {
  /** the object to describe */
  object: PropTypes.any,
};

ObjectValue.contextTypes = {
  theme: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
};

export default ObjectValue;
