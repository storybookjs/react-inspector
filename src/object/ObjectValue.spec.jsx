import React from 'react';
import { render, screen } from '@testing-library/react';
import { ObjectValue } from './ObjectValue';
import { describe, it, expect } from 'vitest';

describe('ObjectValue', () => {
  it('should render', () => {
    const { container } = render(<ObjectValue object={0} />);
    expect(container).toMatchSnapshot();
  });

  it('should render bigint', () => {
    render(<ObjectValue object={9007199254740993n} />);
    const element = screen.getByText('9007199254740993n');
    expect(element.tagName.toLowerCase()).toBe('span');
    expect(element.textContent).toBe('9007199254740993n');
  });

  it('should render number', () => {
    render(<ObjectValue object={0} />);
    const element = screen.getByText('0');
    expect(element.tagName.toLowerCase()).toBe('span');
    expect(element.textContent).toBe('0');
  });

  it('should render string with quotes', () => {
    render(<ObjectValue object={'octocat'} />);
    const element = screen.getByText(/"octocat"/);
    expect(element.tagName.toLowerCase()).toBe('span');
    expect(element.textContent).toBe('"octocat"');
  });

  it('should render boolean', () => {
    render(<ObjectValue object={true} />);
    const element = screen.getByText('true');
    expect(element.tagName.toLowerCase()).toBe('span');
    expect(element.textContent).toBe('true');
  });

  it('should render undefined', () => {
    render(<ObjectValue object={undefined} />);
    const element = screen.getByText('undefined');
    expect(element.tagName.toLowerCase()).toBe('span');
    expect(element.textContent).toBe('undefined');
  });

  it('should render null', () => {
    render(<ObjectValue object={null} />);
    const element = screen.getByText('null');
    expect(element.tagName.toLowerCase()).toBe('span');
    expect(element.textContent).toBe('null');
  });

  it('should display date correctly', () => {
    const dateString = 'December 17, 1995 03:24:00';
    const date = new Date(dateString);
    render(<ObjectValue object={date} />);
    const element = screen.getByText(date.toString());
    expect(element.tagName.toLowerCase()).toBe('span');
    expect(element.textContent).toBe(date.toString());
  });

  it('should render array with length information', () => {
    render(<ObjectValue object={[1, 2, 3, 4, 5]} />);
    const element = screen.getByText('Array(5)');
    expect(element.tagName.toLowerCase()).toBe('span');
    expect(element.textContent).toBe('Array(5)');
  });

  it('should render an empty object', () => {
    render(<ObjectValue object={{}} />);
    const element = screen.getByText('Object');
    expect(element.tagName.toLowerCase()).toBe('span');
    expect(element.textContent).toBe('Object');
  });

  it('should render a simple object', () => {
    render(<ObjectValue object={{ k: 'v' }} />);
    const element = screen.getByText('Object');
    expect(element.tagName.toLowerCase()).toBe('span');
    expect(element.textContent).toBe('Object');
  });

  it('should render a null prototyped object', () => {
    render(<ObjectValue object={Object.create(null)} />);
    const element = screen.getByText('Object');
    expect(element.tagName.toLowerCase()).toBe('span');
    expect(element.textContent).toBe('Object');
  });

  /*
  it('should render an anonymous function', () => {
    renderer.render(<ObjectValue object={function(){}} />)
    const tree = renderer.getRenderOutput()
    expect(tree.type).toBe('span')
    // const n = tree.children[1].props.children[1]
    // console.log(JSON.stringify(n, null, 2))
    // console.log(typeof n)
    // console.log(tree.children[1].props.children)
    expect(tree.children[0]).toEqual(<span style={objectStyles.value.function.keyword}>function</span>)
    expect(tree.children[1]).toEqual(<span style={objectStyles.value.function.name}>{[ '\xa0', 'object', '()' ]}</span>)
  });
  */

  /*
  it('should render a named function', () => {
    renderer.render(<ObjectValue object={function id(a){return a;}} />)
    const tree = renderer.getRenderOutput()
    expect(tree.type).toBe('span')
    expect(tree.children).toEqual([
                                          <span style={objectStyles.value.function.keyword}>function</span>,
                                          <span style={objectStyles.value.function.name}>{[ '\xa0', 'id', '()' ]}</span>
                                        ])
  });
  */

  it('should render a symbol', () => {
    render(<ObjectValue object={Symbol()} />);
    const element = screen.getByText('Symbol()');
    expect(element.tagName.toLowerCase()).toBe('span');
    expect(element.textContent).toBe('Symbol()');
  });

  it('should render a symbol foo', () => {
    render(<ObjectValue object={Symbol('foo')} />);
    const element = screen.getByText('Symbol(foo)');
    expect(element.tagName.toLowerCase()).toBe('span');
    expect(element.textContent).toBe('Symbol(foo)');
  });

  it('accepts and applies style from `styles` prop', () => {
    // Custom `styles` prop gets applied to the element
    const style = { color: 'blue' };
    render(<ObjectValue styles={style} object={''} />);
    const element = screen.getByText('""');

    expect(element.getAttribute('style')).toMatchInlineSnapshot(`
      "color: blue;"
    `);
  });
});
