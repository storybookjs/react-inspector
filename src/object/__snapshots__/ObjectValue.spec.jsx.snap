// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ObjectValue > should render 1`] = `
<div>
  <span
    style="color: rgb(28, 0, 207);"
  >
    0
  </span>
</div>
`;
